primo2/inference/dynamic.py

```diff
@@ -0,0 +1,305 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+# This file is part of PRIMO2 -- Probabilistic Inference Modules.
+# Copyright (C) 2013-2017 Social Cognitive Systems Group, 
+#                         Faculty of Technology, Bielefeld University
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the Lesser GNU General Public License as 
+# published by the Free Software Foundation, either version 3 of the 
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public 
+# License along with this program.  If not, see
+# <http://www.gnu.org/licenses/>.
+
+import abc
+import copy
+import six
+
+from . import exact
+
+
+__all__ = [
+    'PriorFeedbackExact',
+    'SoftEvidenceExact'
+]
+
+
+@six.add_metaclass(abc.ABCMeta)
+class DBNInferenceMethod(object):
+    """Abstract class for dynamic Bayesian network inference methods.
+
+    Contains the inference algorithm and the current state of the 
+    network at time t.
+    """
+
+    def __init__(self, dbn):
+        """Create the inference method and its state.
+
+        Parameters
+        ----------
+        dbn : DynamicBayesianNetwork
+            A 2-TBN dynamic Bayesian network structure.
+        """
+        super(DBNInferenceMethod, self).__init__()
+        self._dbn = dbn
+        self._t = 0
+        self._evidence = False
+
+    @abc.abstractmethod
+    def marginal_probabilities(self, variable):
+        """Get the marginal probabilities of a variable. (abstract)
+
+        Parameters
+        ----------
+        variable : RandomNode|String
+            The variable/node.
+
+        Returns
+        -------
+        dict
+            A dictionary, dict-keys are the values of the variable/node and 
+            dict-values are the probabilities of each value.
+        """
+        pass
+
+    @abc.abstractmethod
+    def marginals(self, variables):
+        """Get the marginals of a variable. (abstract)
+
+        Parameters
+        ----------
+        variables : [String,]
+            List containing the names of the variables whose joint prior
+            or posterior marginals are desired.   
+                
+        Returns
+        -------
+        ?
```

Comment:
Can you think of a sensible alternative? I mean in theory one could do inference solely on the cpts but that would kind of go against the point of splitting specification and calculation (which granted we already violate a little bit with the sampling approaches).